lib/engine/game/base.rb

```diff
@@ -1840,22 +1840,26 @@ def buy_train(operator, train, price = nil)
       end
 
       def discountable_trains_for(corporation)
-        discountable_trains = @depot.depot_trains.select(&:discount)
+        discountable_trains = @depot.depot_trains.select { |train| train.discount || train.variants.any? { |_, v| v[:discount] } }
 
         corporation.trains.flat_map do |train|
           discountable_trains.flat_map do |discount_train|
+            discount_info = []
             discounted_price = discount_train.price(train)
-            next if discount_train.price == discounted_price
-
-            name = discount_train.name
-            discount_info = [[train, discount_train, name, discounted_price]]
+            if discount_train.price > discounted_price
+              name = discount_train.name
+              discount_info = [[train, discount_train, name, discounted_price]]
```

Comment:
remove name = discount_train.name and use discount_train.name directly since its only used once

Reply:
it's used bellow, L1856 as well